Add `isort` dependency and pre-commit hook

harripj opened the pull request:
#### Description of the change
As discussed with @hakonanes and @din14970 in https://github.com/pyxem/orix/pull/334#discussion_r873160646, `isort` is a package that may be used to automatically sort imports. It also hooks into pre-commit to allow for formatting before code submission.

We could consider adding `isort` as a dependency to `orix` to handle import formatting. In `pre-commit-config.yaml` I have added the [standard pre-commit config](https://pycqa.github.io/isort/docs/configuration/pre-commit.html) along with the `black` profile and also with the `"--force-sort-within-sections"` [flag](https://pycqa.github.io/isort/docs/configuration/options.html#sort-relative-in-force-sorted-sections) which groups imports by modules (as we currently do) rather than sorting straight `import`s before `from`-style imports. 


#### Progress of the PR
- [n/a] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [n/a] Unit tests with pytest for all lines
- [x] Clean code style by [running black via pre-commit](https://orix.readthedocs.io/en/latest/contributing.html#code-style)

#### For reviewers
<!-- Don't remove the checklist below. -->
- [ ] The PR title is short, concise, and will make sense 1 year later.
- [ ] New functions are imported in corresponding `__init__.py`.
- [ ] New features, API changes, and deprecations are mentioned in the unreleased
      section in `CHANGELOG.rst`.
- [ ] Contributor(s) are listed correctly in `__credits__` in `orix/__init__.py` and in
      `.zenodo.json`.

hakonanes commented on CONTRIBUTING.rst:
```diff
@@ -75,6 +76,11 @@ Note that ``black`` won't format `docstrings
+used to prevent sorting.
+
 Comment lines should preferably be limited to 72 characters.
 
```
The paragraph below this line on structuring on package imports is redundant now, right? Or, a note that we're using `isort` could be added as an extra line to the paragraph.

hakonanes commented on orix/plot/stereographic_plot.py:
```diff
@@ -33,17 +33,20 @@
 import numpy as np
 from scipy.ndimage import gaussian_filter
 
+# fmt: off
```
Is it necessary to also tell Black to not format these lines? Black didn't have to be told before, so why now?

harripj replied:
Annoyingly `black` forces a new line before a comment, which I agree is a good idea but not for these action comments as it actually breaks up imports that may be from the same module. The alternative is:

```
# isort: off
import a.c
import a.b

# isort: on
import a.d, others...
```

Happy with either to be honest, perhaps it is easiest just to accept the new line (ie. as above), what do you think?

hakonanes replied:
Hm, I'd say leave the Black comments in this case. Annoying, but worth it.

hakonanes commented on orix/tests/quaternion/test_orientation.py:
```diff
@@ -30,9 +30,9 @@
-    T,
     O,
     Oh,
+    T,
```
Point groups are listed from lowest to highest symmetry, so this list should get a `#isort: skip` flag, right?

hakonanes replied:
Or `off/on`.

hakonanes commented on orix/tests/sampling/test_cubochoric_sampling.py:
```diff
@@ -20,12 +20,12 @@
 
 from orix.quaternion.rotation import Rotation
-from orix.quaternion.symmetry import C1, C2, D2, C4, D4, C3, D3, C6, D6, T, O
+from orix.quaternion.symmetry import C1, C2, C3, C4, C6, D2, D3, D4, D6, O, T
```
Shouldn't be sorted by isort.

hakonanes commented on orix/vector/__init__.py:
```diff
@@ -25,10 +25,12 @@
 
-# Must be imported first
-from orix.vector.vector3d import check_vector, Vector3d
+# fmt: off
```
Again, do we have to tell both isort and Black to not sort?

hakonanes reviewed (changes requested):
This makes the imports easier and faster to read and frees up review time to be used on implementation and not code style, good addition!

Is it possible to add an `isort` check to the "code style" CI workflow?

harripj commented:
Should be okay for another look @hakonanes.

The blank line comment behaviour discussed above is still present, despite using [certain parameters](https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html#custom-configuration). If `isort: skip` is used then imports will actually sort around the skipped line (not what we want). `isort: split` is better but as far as I can work out will add a new line around the split section. So I ended up using `fmt` and `isort` `on/off` around the regions, which gives the correct behaviour.

Note I also updated the `black-jupyter` CI test by adding the `--check` option. With an ill-formatted notebook previously the [test would pass](https://github.com/harripj/orix/runs/6690792272?check_suite_focus=true#step:7:11) but actually reformat the notebook, but now the [test will fail](https://github.com/pyxem/orix/runs/6690830763?check_suite_focus=true#step:7:12) which I think is more inline with what we want to test.

hakonanes commented on .github/workflows/build.yml:
```diff
@@ -9,12 +9,19 @@ jobs:
+          python-version: 3.9
+      - uses: isort/isort-action@master
+        with:
+          configuration: --profile black --ff --fss --check-only --diff
```
Could you write out the options here? `--from-first` instead of `--ff` and `--force-sort-within-sections` instead of `--fss`. So we don't have to look them up to remember what they mean.

harripj replied:
Good idea, in fact the `--ff` flag was incorrect, I have corrected it to `--filter-files`. Thanks!

hakonanes reviewed (approved):
Minor comment which isn't strictly necessary but improves readability.

> So I ended up using fmt and isort on/off around the regions, which gives the correct behaviour.

I think that's fine. Thank you for checking.

> Note I also updated the black-jupyter CI test by adding the --check option

Yes, thank you!

hakonanes commented:
Thanks! Will merge once checks pass.
